evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
+			err = fmt.Errorf("panic in processing message: %v", e)
+		}
+	}()
+
+	switch chID {
+	case EvidenceChannel:
+		err = r.handleEvidenceMessage(envelope)
+
+	default:
+		err = fmt.Errorf("unknown channel ID (%d) for envelope (%v)", chID, envelope)
+	}
+
+	return err
+}
+
+// processEvidenceCh implements a blocking event loop where we listen for p2p
+// Envelope messages from the evidenceCh.
+func (r *Reactor) processEvidenceCh() {
+	defer r.evidenceCh.Close()
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
+				r.evidenceCh.Error() <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
+				}
+			}
+
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on evidence channel; closing...")
 			return
-		case nil:
-		default:
-			// continue to the next piece of evidence
-			evR.Logger.Error("Evidence has not been added", "evidence", evis, "err", err)
 		}
 	}
 }
 
-// SetEventBus implements events.Eventable.
-func (evR *Reactor) SetEventBus(b *types.EventBus) {
-	evR.eventBus = b
+// processPeerUpdate processes a PeerUpdate, returning an error upon failing to
+// handle the PeerUpdate or if a panic is recovered. For new or live peers it
+// will check if an evidence broadcasting goroutine needs to be started. For
+// down or removed peers, it will check if an evidence broadcasting goroutine
+// exists and signal that it should exit.
+//
+// FIXME: The peer may be behind in which case it would simply ignore the
+// evidence and treat it as invalid. This would cause the peer to disconnect.
+// The peer may also receive the same piece of evidence multiple times if it
+// connects/disconnects frequently from the broadcasting peer(s).
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("panic in processing peer update: %v", e)
+		}
+	}()
+
+	r.Logger.Debug("received peer update", "peer", peerUpdate.PeerID.String(), "status", peerUpdate.Status)
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Check if we've already started a goroutine for this peer, if not we create
+		// a new done channel so we can explicitly close the goroutine if the peer
+		// is later removed, we increment the waitgroup so the reactor can stop
+		// safely, and finally start the goroutine to broadcast evidence to that peer.
+		_, ok := r.peerRoutines[peerUpdate.PeerID.String()]
+		if !ok {
+			r.peerRoutines[peerUpdate.PeerID.String()] = make(chan struct{})
+			r.peerWG.Add(1)
+			go r.broadcastEvidenceLoop(peerUpdate.PeerID, r.peerRoutines[peerUpdate.PeerID.String()])
```

Comment:
Maybe use a local `doneCh` variable to make it clearer what the second argument is.

Reply:
Sure thing!